code/datums/components/food/ice_cream_holder.dm

```diff
@@ -0,0 +1,280 @@
+	else
+		to_chat(user, "<span class='warning'>[source] can't hold anymore ice cream!</span>")
+	return COMPONENT_CANCEL_ATTACK_CHAIN
+
+/// Adds more ice cream.
+/datum/component/ice_cream_holder/proc/add_serving(atom/source, flavour_name, list/reagents, custom_name, forced = FALSE)
+	var/datum/ice_cream_flavour/flavour = GLOB.ice_cream_flavours[flavour_name]
+	return (forced || length(scoops) < max_scoops) && flavour.add_flavour(src, reagents, custom_name)
+
+/////ICE CREAM FLAVOUR DATUM STUFF
+
+GLOBAL_LIST_INIT_TYPED(ice_cream_flavours, /datum/ice_cream_flavour, init_ice_cream_flavours())
+
+/proc/init_ice_cream_flavours()
+	. = list()
+	for(var/datum/ice_cream_flavour/flavour as anything in subtypesof(/datum/ice_cream_flavour))
+		flavour = new flavour
+		.[flavour.name] = flavour
+
+/*
+ * The ice cream flavour datums. What makes these digital frozen snacks so yummy.
+ * They are singletons, so please bear with me if they feel a little tortous to use at time.
+ */
+/datum/ice_cream_flavour
+	/// Make sure the same name is not found on other types; These are singletons keyed by their name after all.
+	var/name = "Coderlicious Gourmet Double Deluxe Undefined"
+	/// The icon state of the flavour, overlay or not.
+	var/icon_state = "icecream_vanilla"
+	/*
+	 * The fluff text sent to the examiner when the snack has only one flavour of ice cream. Make sure your new subtypes have one.
+	 * $CONE_NAME and $CUSTOM_NAME are both placeholder names for the cone and the custom ice cream name respectively.
+	 */
+	var/desc = ""
+	/*
+	 * Depending on [/datum/component/ice_cream/var/change_desc] bool value, 'desc' may effectively be the description or shown
+	 * on [/atom/proc/examine_more]. In the former case, the desc is joined with prefix.
+	 */
+	var/desc_prefix = "A delicious $CONE_NAME"
+	/// The ingredients required to produce a unit, these are multiplied by 3.
+	var/list/ingredients = list(/datum/reagent/consumable/milk, /datum/reagent/consumable/ice, /datum/reagent/consumable/vanilla)
+	/// The same as above, but in a readable text generated on spawn that can also contain extra ingredients such as "lot of love" or "optional flavorings".
+	var/ingredients_text = ""
+	/// reagent added in 'add_flavour'
+	var/reagent_type
+	/// the amount of reagent added in 'add_flavour'
+	var/reagent_amount = 3
+	/// Is this shown on the ice cream vat menu or not?
+	var/hidden = FALSE
+
+/datum/ice_cream_flavour/New()
+	if(ingredients)
+		ingredients_text = "(Ingredients: [reagent_paths_list_to_text(ingredients, ingredients_text)])"
+
+///Adds a new flavour to the ice cream cone.
+/datum/ice_cream_flavour/proc/add_flavour(datum/component/ice_cream_holder/target, datum/reagents/R, custom_name)
+	var/atom/owner = target.parent
+	LAZYADD(target.scoops, custom_name || name)
+	if(icon_state)
+		LAZYADD(target.scoop_overlays, icon_state)
+	if(custom_name)
+		LAZYSET(target.special_scoops, custom_name, name)
+
+	owner.reagents.maximum_volume += EXTRA_MAX_VOLUME_PER_SCOOP
```

Comment:
This looks abusable if the edible is extractable ?

Reply:
Slightly, but most edibles are not extractable via syringe, plus that would count more as a creative usage of a food container than else (it's probably not a big deal, but i don't want to jinx it).